src/controllers/cartController.js

```diff
@@ -0,0 +1,180 @@
+const Cart = require('../models/Cart');
+const Book = require('../models/Book');
+const { calculateCartTotal } = require('../utils/cartTotal');
+const stripe = require('stripe')(process.env.STRIPE_SECRET_KEY);
+const { NotFoundError, BadRequestError } = require('../errors');
+const { StatusCodes } = require('http-status-codes');
+
+// get the cart
+const getCart = async (req, res, next) => {
+  try {
+    const userId = req.user.userId;
+
+    const cart = await Cart.findOne({ createdBy: userId }).populate(
+      'orderItems.book'
+    );
+    if (!cart) {
+      throw new NotFoundError('Cart not found');
+    }
+
+    const updatedItems = cart.orderItems.filter(
+      (item) => item.book && item.book.isAvailable
+    );
+
+    if (updatedItems.length !== cart.orderItems.length) {
+      cart.orderItems = updatedItems;
+      cart.total = calculateCartTotal(cart);
+
+      await cart.save();
+
+      return res.status(StatusCodes.OK).json({
+        message: 'Some unavailable books were removed from your cart.',
+        cart,
+      });
+    }
+
+    res.status(StatusCodes.OK).json({ cart });
+  } catch (error) {
+    next(error);
+  }
+};
+
+// add book to catr
+const addToCart = async (req, res, next) => {
+  try {
+    const userId = req.user.userId;
+    const { bookId } = req.body;
+
+    const book = await Book.findById(bookId);
+    if (!book || !book.isAvailable) {
+      throw new BadRequestError('Book is unavailable or does not exist');
+    }
+
+    let cart = await Cart.findOne({ createdBy: userId });
```

Comment:
This can cause a race condition if 2 users try to add the same item to cart, we should use atomic updates, with findOneAndUpdate: 
 { createdBy: userId },
  {
    $addToSet: {
      orderItems: {
        book: book._id,
        title: book.title,
        author: book.author,
        coverImageUrl: book.coverImageUrl,
        price: book.price,
        isAvailable: book.isAvailable,
      },
    },
  },
  { upsert: true, new: true }
);

We covered this in our first meeting, so I doubt it still makes sense(I probably explained it with a vending machine)

Reply:
Thank you! Got it!